DEP: Do not allow "abstract" dtype conversion/creation

seberg opened the pull request:
These dtypes do not really make sense as instances. We can (somewhat)
reasonably define np.dtype(np.int64) as the default (machine endianess)
int64. (Arguably, it is unclear that `np.array(arr_of_>f8, dtype="f")`
should return arr_of_<f8, but that would be very noisy!)

However, `np.integer` as equivalent to long, is not well defined.
Similarly, `dtype=Decimal` may be neat to spell `dtype=object` when you
intend to put Decimal objects into the array. But it is misleading,
since there is no special meaning to it at this time.

The biggest issue with it, is that `arr.astype(np.floating)` looks
like it will let float32 or float128 pass, but it will force a
float64 output! Arguably downcasting is a bug in this case.

A related issue is `np.dtype("S")` and especially "S0". The dtype "S"
does make sense for most or all places where `dtype=...` can be
passed. However, it is conceptionally different from other dtypes, since
it will not end up being attached to the array (unlike "S2" which
would be). The dtype "S" really means the type number/DType class
of String, and not a specific dtype instance.

---

I have a pretty strong opinion, that at least these super classes should be deprecated. I can be waved about the `object` case, although it seems very confusing to accept it.
The conversion of `S0`, and `np.dtype("S")` is another point, which I do not accept as correct, but requires a new converter with special handling for `"S"` and `np.str_` to fix. (Honestly, I think the above example of `np.array(np.array(..., dtype=">f8"), dtype=np.float64).dtype == "<float64"` is unclear to be correct logic, but... it is convenient logic)

seberg committed:
DEP: Do not allow "abstract" dtype conversion/creation

These dtypes do not really make sense as instances. We can (somewhat)
reasonably define np.dtype(np.int64) as the default (machine endianess)
int64. (Arguably, it is unclear that `np.array(arr_of_>f8, dtype="f")`
should return arr_of_<f8, but that would be very noisy!)

However, `np.integer` as equivalent to long, is not well defined.
Similarly, `dtype=Decimal` may be neat to spell `dtype=object` when you
intend to put Decimal objects into the array. But it is misleading,
since there is no special meaning to it at this time.

The biggest issue with it, is that `arr.astype(np.floating)` looks
like it will let float32 or float128 pass, but it will force a
float64 output! Arguably downcasting is a bug in this case.

A related issue is `np.dtype("S")` and especially "S0". The dtype "S"
does make sense for most or all places where `dtype=...` can be
passed. However, it is conceptionally different from other dtypes, since
it will not end up being attached to the array (unlike "S2" which
would be). The dtype "S" really means the type number/DType class
of String, and not a specific dtype instance.

seberg commented on numpy/lib/_iotools.py:
```diff
@@ -504,18 +504,23 @@ class StringConverter:
+                    # ones (should only be used for the converter)
+                    (nx.integer, int, -1),
+                    (nx.floating, float, nx.nan),
+                    (nx.complexfloating, complex, nx.nan + 0j),])
```
This is the most annoying part of the change. I think it is OK, basically... This only gets called when the dtype is was already fixed, in which case the dtype is not used, but only the converters is found.

eric-wieser replied:
Turns out this is not ok:

This change has broken `genfromtxt` with dtype object:

Before:
```python
>>> np.genfromtxt(io.StringIO("1"), dtype=object)
array(b'1', dtype=object)
>>> np.genfromtxt(io.StringIO("what"), dtype=object)
array(b'what', dtype=object)
```

After:
```python
>>> np.genfromtxt(io.StringIO("1"), dtype=object)
array((1+0j), dtype=object)
>>> np.genfromtxt(io.StringIO("what"), dtype=object)
array(None, dtype=object)
```

seberg replied:
Ouch, subtle, I guess string being the last entry is important, probably switching the order will be OK, but have to look...

eric-wieser replied:
Filed as #16189

seberg commented:
Just to note, I checked scipy for `np.integer` occurance, and it looks like at most one or two things where it is used incorrectly. NumPy had more. A few of the "object" changes, are arguably unintended, the author probably expected to get an array of `Decimal` not integer objects. So when it comes to libraries, I actually do not expect it to be painful at all. And when it comes to users: We do not have to change it tomorrow...

eric-wieser commented on numpy/core/src/multiarray/descriptor.c:
```diff
@@ -1425,6 +1436,12 @@ _convert_from_type(PyObject *obj) {
 
+        if (DEPRECATE("Converting a type/class not known to NumPy to a dtype "
+                      "currently always returns `np.dtype(object)`. This loses "
+                      "the type information and is deprecated.") < 0) {
```
I'm inclined not to deprecate this - it's a way to optimistically write code in a way that assumes maybe numpy will natively support the type in the distant future, and do "the right thing".

seberg replied:
Yeah, I was not quite sure about this one, the optimistically writing does make a lot of sense (it is also a bit of a trap), but...
If we never deprecate this, it can eventually lead to subtle behaviour changes in downstream code. Usually, to the *intended* behaviour hopefully.

I guess I will remove this part, and see if I can add a note to the dtype documentation.

I somewhat intend on deprecating `np.dtype("S")` (still allowing it for most `dtype="S"` arguments). At that point, I will probably revisit this part.

seberg committed:
Fixup: Do not deprecate generic python types

Add a comment that we can think about only allowing it for `dtype=...`
though...

seberg commented:
We discussed this briefly today, and nobody seemed concerned. Also the mailing list discussion did not bring up any serious concerns about the warnings. And I do think we have general agreement that this is the correct behaviour.

mattip commented on numpy/core/src/multiarray/descriptor.c:
```diff
@@ -1425,7 +1436,13 @@ _convert_from_type(PyObject *obj) {
+        /*
+         * All other classes are treated as object. This can be convenient
+         * to convey an intention of using it for a specific python type
+         * and possibly allow future enforcement. It may make sense to
```
```suggestion
         * and possibly allow converting to a new type-specific dtype in the future. It may make sense to
```
maybe?

mattip commented:
LGTM. A little nit about wording in a c-comment.

seberg committed:
Update numpy/core/src/multiarray/descriptor.c

Co-Authored-By: Matti Picus <matti.picus@gmail.com>

seberg commented:
@mattip restarting it, the PyPy tests seem to fail fairly often right now? This time in:
```
numpy/core/tests/test_einsum.py ...........Fatal Python error: Illegal instruction
```

mattip commented:
Thanks @seberg. I filed an [issue](https://foss.heptapod.net/pypy/pypy/issues/3186) with PyPy
